stdpopsim/simulation_engines.py

```diff
@@ -0,0 +1,92 @@
+import logging
+
+import attr
+import msprime
+import stdpopsim
+
+logger = logging.getLogger(__name__)
+
+_registered_simulation_engines = {}
+
+
+def register_simulation_engine(engine):
+    """
+    Registers the specified simulation engine.
+    """
+    logger.debug(f"Registering simulation engine '{engine.id}'")
+    _registered_simulation_engines[engine.id] = engine
+
+
+def get_simulation_engine(id):
+    """
+    Returns the simulation engine with the specified id.
+    """
+    if id not in _registered_simulation_engines:
+        raise ValueError(f"Simulation engine '{id}' not registered")
+    return _registered_simulation_engines[id]
+
+
+def all_simulation_engines():
+    """
+    Returns an iterator over all registered simulation engines.
+    """
+    for sim_engine in _registered_simulation_engines.values():
+        yield sim_engine
+
+
+@attr.s(frozen=True)
+class SimulationEngine(object):
+    """
+    Class representing a simulation engine.
+
+    :ivar id The unique identifier for the simulation engine.
+    :vartype id: str
+    :ivar name: The informal name for this engine as it would be used in
+        written text.
+    :vartype name: str
+    :ivar citations: A list of :class:`.Citation` objects for the simulation
+        engine.
+    :vartype citations: list
+    :ivar add_arguments: Function accepting an :class:`argparse.ArgumentParser`
+        and defining engine specific command line arguments. Once parsed,
+        these will be passed to the `simulate` function kwargs. May be `None`.
+    :ivar simulate: Function to execute the acutal simulation.
+        See msprime example in `simulation_engines.py`.
+    """
+    id = attr.ib(type=str, kw_only=True)
+    name = attr.ib(type=str, kw_only=True)
+    citations = attr.ib(factory=list, kw_only=True)
+    add_arguments = attr.ib(default=None, kw_only=True)  # function
+    simulate = attr.ib(kw_only=True)  # function
+
+
+def _msprime_simulate_wrapper(model, contig, samples, **kwargs):
```

Comment:
What we could do here is,
```python
def _msprime_simulate_wrapper(model=None, contig=None, samples=None, *, **kwargs):
```
This would allow us to use the initial three args as keyword args, and also make sure that all subsequence args are keywords. (Second bit isn't essential)

Reply:
I think we want the`SimulationEngine` class to carry around the version number of the engine too. Can we add a class variable for that? Also is that simple enough to access from SLiM programmatically? I don't know the SLiM side well

Reply:
So, you want to call out to the slim binary each time the stdpopsim library is imported to populate this variable? Or did you mean you want a minimum supported version number?

I don't think there's a simple function to get the version number from within SLiM. If you were really keen, of course, you could do something crazy like: `system("slim -version | awk '{split($3, a, ","); print a[1]}' ")` and capture the output.

Reply:
yes i was thinking the former-- calling out to `slim` to get the version, if it were not available from returned tree sequence. that simple `system()` call is absolutely fine by me, if not elegant.

Reply:
``get_version()`` can be a method of the engine, called on demand. I think the version information is captured in detail in the Provenance information in the returned tree sequence too though.

Reply:
I added a get_version() method (its simpler than using an attribute for calling to an external binary), and changed changed engine implementation to be via subclassing. The attr.ib approach seems inappropriate for classes containing many methods.